src/storage/mod.rs

```diff
@@ -0,0 +1,593 @@
+}
+
+impl StorageSubscription {
+    pub fn new<
+        S: StorageBacking + StorageSubscriber<S>,
+        T: DeserializeOwned + Send + Sync + Clone + 'static,
+    >(
+        tx: Sender<StorageChannelPayload>,
+        key: S::Key,
+    ) -> Self {
+        let getter = move || {
+            let data = S::get::<T>(&key).unwrap();
+            StorageChannelPayload::new(data)
+        };
+        Self {
+            getter: Box::new(getter),
+            tx: Arc::new(tx),
+        }
+    }
+
+    /// Gets the latest data from storage and sends it to the channel.
+    pub fn get_and_send(&self) -> Result<(), SendError<StorageChannelPayload>> {
+        let payload = (self.getter)();
+        self.tx.send(payload)
+    }
+}
+
+// End StorageSenderEntry
+
+// Start StorageChannelPayload
+
+/// A payload for a storage channel that contains the latest value from storage.
+#[derive(Clone, Debug)]
+pub struct StorageChannelPayload {
+    data: Arc<dyn Any + Send + Sync>,
+}
+
+impl StorageChannelPayload {
+    /// Creates a new StorageChannelPayload
+    pub fn new<T: Send + Sync + 'static>(data: T) -> Self {
+        Self {
+            data: Arc::new(data),
+        }
+    }
+
+    /// Gets the data from the payload
+    pub fn data<T: 'static>(&self) -> Option<&T> {
+        self.data.downcast_ref::<T>()
+    }
+}
+
+impl Default for StorageChannelPayload {
+    fn default() -> Self {
+        Self { data: Arc::new(()) }
+    }
+}
+// End StorageChannelPayload
+
+// Start helper functions
+
+/// Serializes a value to a string and compresses it.
+pub(crate) fn serde_to_string<T: Serialize>(value: &T) -> String {
+    let serialized = to_allocvec(value).unwrap();
+    let compressed = yazi::compress(
+        &serialized,
+        yazi::Format::Zlib,
+        yazi::CompressionLevel::BestSize,
+    )
+    .unwrap();
+    let as_str: String = compressed
+        .iter()
+        .flat_map(|u| {
+            [
+                char::from_digit(((*u & 0xF0) >> 4).into(), 16).unwrap(),
+                char::from_digit((*u & 0x0F).into(), 16).unwrap(),
+            ]
+            .into_iter()
+        })
+        .collect();
+    as_str
+}
+
+#[allow(unused)]
+/// Deserializes a value from a string and unwraps errors.
+pub(crate) fn serde_from_string<T: DeserializeOwned>(value: &str) -> T {
+    try_serde_from_string(value).unwrap()
+}
+
+/// Deserializes and decompresses a value from a string and returns None if there is an error.
+pub(crate) fn try_serde_from_string<T: DeserializeOwned>(value: &str) -> Option<T> {
+    let mut bytes: Vec<u8> = Vec::new();
+    let mut chars = value.chars();
+    while let Some(c) = chars.next() {
+        let n1 = c.to_digit(16)?;
+        let c2 = chars.next()?;
+        let n2 = c2.to_digit(16)?;
+        bytes.push((n1 * 16 + n2) as u8);
+    }
+    let (decompressed, _) = yazi::decompress(&bytes, yazi::Format::Zlib).ok()?;
+    postcard::from_bytes(&decompressed).ok()
```

Comment:
This is throwing [`WontImplement`](https://docs.rs/postcard/latest/postcard/enum.Error.html#variant.WontImplement), and the error is ignored due to `ok()`, which causes the issue I mentioned [above in the comments](https://github.com/DioxusLabs/dioxus-std/pull/17#issuecomment-1800687581).

I'm just using `new_storage` on M1 macOS.

---

FTR, I rewrote this to print the errors (perhaps these errors should be logged instead of ignored?):

```rust
    match yazi::decompress(&bytes, yazi::Format::Zlib) {
        Ok((decompressed, _)) => match postcard::from_bytes(&decompressed) {
            Ok(v) => Some(v),
            Err(err) => {
                println!("ERR(from_bytes): {:?}", err);
                None
            }
        },
        Err(err) => {
            println!("ERR(decompress): {:?}", err);
            None
        }
    }
```

Reply:
Aside from the use of `BTreeMap`, there's nothing special in my signal structs. Is there a reason `postcard` was chosen? It has trouble deserializing in certain cases (throwing `WontImplement`):

- https://github.com/jamesmunns/postcard/issues/77
- https://github.com/jamesmunns/postcard/issues/79

Reply:
> Is there a reason postcard was chosen?

Yes, it doesn't add very much size to the binary size, it is a binary format so it serialized into a smaller string, and it is cross platform

Reply:
The errors are now logged

Reply:
This PR is getting quite large, so I don't think it should be in this PR, but in the future we could support other serialization formats with a cfg flag or generic

Reply:
> The errors are now logged

Can confirm that it comes as ERROR (my project uses `tracing`):

<img width="603" alt="image" src="https://github.com/DioxusLabs/dioxus-std/assets/3998/daf278e3-7b17-4dea-9ed2-79e43011b29a">

Reply:
> This PR is getting quite large, so I don't think it should be in this PR, but in the future we could support other serialization formats with cfg flags or a new generic

I've opened https://github.com/DioxusLabs/dioxus-std/issues/20 for issues related in `postcard` in particular.